### PR TITLE
Unextracted all cabal hashes

### PR DESCRIPTION
**Pull request by ElvishJerricco**
###### Motivation for this change

#30290, while useful, is kind of massively complicated, and actually leads to hashes which are different depending on the OS. We can get rid of the crazy C program and special schema, *and* reduce disk usage if we just leave all the files in the tarball and extract them as needed to their own store paths. This produces unique derivations and therefore unique paths, even given macOS's case insensitivity, without having to conjure those paths manually.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by ElvishJerricco** [comment]
cc: @shlevy

**Review by shlevy — APPROVED** [review]
:fireworks:

**Comment by shlevy** [comment]
Barring objections from @peti, ping me in a few days and I'll merge

**Review by peti — APPROVED** [review]
👍

File: pkgs/data/misc/hackage/default.nix [hunk 1]
```diff
@@ -1,34 +1,6 @@
...
-        installPhase =
-          ''
-            mkdir -p $out/bin
-            install -m755 partition-all-cabal-hashes $out/bin
-          '';
-      };
-in fetchFromGitHub (src-spec //
-  { postFetch = "${partition-all-cabal-hashes}/bin/partition-all-cabal-hashes $downloadedFile $out";
-  })
+fetchurl {
```
Comment (NeQuissimus) [inline pkgs/data/misc/hackage/default.nix]: Why did you prefer `fetchurl` over `fetchFromGitHub` here?

Comment (ElvishJerricco) [reply pkgs/data/misc/hackage/default.nix]: `fetchFromGitHub` is the whole cause of the problem. It extracts the tarball in full to a directory, which breaks on macOS when that tarball has paths that clash on a case insensitive file system.

Comment (NeQuissimus) [reply pkgs/data/misc/hackage/default.nix]: Ah, I got it! This whole ordeal does make sense now :D

**Comment by Ericson2314** [comment]
Should we remove that extra hook from fetchurl / fetchgit too?

**Comment by mightybyte** [comment]
@shlevy Any chance we can get this back ported to 17.03?

**Comment by shlevy** [comment]
Sure, ping me on the PR

**Comment by ElvishJerricco** [comment]
@shlevy Should I just open a PR to the 17.03 branch on nixpkgs-channels?

**Comment by shlevy** [comment]
No, the release-17.03 branch here.

**Comment by ElvishJerricco** [comment]
Gotcha.

**Pull request merged**